### ISSUE TITLE
Add per-window-name global settings

### ISSUE THREAD

**Issue by zadjii-msft** [microsoft/terminal#9992]
This one's from a team sync chat thread.

What if there was a way to say

```json
"windows": {
  "work": {
    "alwaysOnTop" : true,
    "initialPosition": "2000, 50",
    "centerOnLaunch": true,
    "launchMode": "focus",
    "startupActions": "sp ; sp ; nt -p Ubuntu"
  },
  "_quake": {
    "launchMode": "default",
    "startupActions": "nt -p PowerShell"
  }
}
```

So that you could have different sets of global settings for different window names? What properties would make sense here? Definitely not `windowingBehavior`, because that's a global setting, not a window setting.

* `dev/migrie/fhl-fall-2023/9992-window-name-settings` https://github.com/microsoft/terminal/compare/main...dev/migrie/fhl-fall-2023/9992-window-name-settings
  * ~`dev/migrie/fhl-fall-2023/9992-default-quake-settings`~ this was a dead end
  * `dev/migrie/fhl-fall-2023/9992-quake-II` https://github.com/microsoft/terminal/compare/dev/migrie/fhl-fall-2023/9992-window-name-settings...dev/migrie/fhl-fall-2023/9992-quake-II
    * `dev/migrie/fhl-fall-2023/11162-quake-III-arena` https://github.com/microsoft/terminal/compare/dev/migrie/fhl-fall-2023/9992-quake-II...dev/migrie/fhl-fall-2023/11162-quake-III-arena for #11162, #11174

------

Edit 2026-march-09:

New year, new hackathon. Updated branch is here:
https://github.com/microsoft/terminal/compare/main...dev/migrie/fhl-spring-2026/quake-4
(I just used one branch for everything - doing docking as a separate branch doesn't make sense frankly)

I think that supports docking, different window names, default window settings + per-window settings, and a lot of hot-reloading for window settings (but not everything). 

A big issue that remains with this branch: there is _absolutely no_ UI for these settings, other than the JSON file. There's still too many inter-mingled global & per-window settings such that it doesn't make sense to have like... a dropdown on the settings pages that tells you which window you're modifying the settings for. That's probably a solvable issue, but surely not during a hackathon. 

Other notes in https://github.com/microsoft/terminal/issues/9992#issuecomment-1719994413

Anyone is free to get this across the finish line - I sure don't have time for it.

**Comment by DHowett** [comment]
1. This would make it so much easier for people to have "work profiles" they come into in the morning
2. Specialization for quake that isn't special
3. GlobalAppSettings is already inheritable (!)

Bad points:

1. what if you rename a window

**Comment by dag03tsc** [comment]
I agree this could be pretty useful indeed. Meanwhile, if anyone needs it, I've managed to alter the behavior of the `_quake` window with this setting:

`"startupActions": "wt -w _quake sp -H -p CMD -s .666 \\; sp -H -p UBUNTU -s .5"`

The above is just an example of a "quaked" 3-panel symmetric vertical arrange. It's a "hackity-hack" bypass, 'cause I'm just using a legit functionality for something it wasn't designed to, but I'll try to explain its two key points.

First point, if `wt -w _quake` is the first subcommand provided, it replaces the initially launched window with a "Quake Mode" enabled one; otherwise, it opens an independent one.

Second point, there's a huge difference between writing semicolons ` ; ` or escaped semicolons ` \\; `, and the arranging of commands is very important; all subcommands before `wt -w _quake` are applied within the initially launched window and must be preceded by unescaped semicolons ` ; `, while all subcommands after `wt -w _quake` are applied within the new "quaked" window and must be preceded by escaped semicolons ` \\; ` in order to bypass the initially launched window parsing.

As an example, let's say you want to launch two separate windows with a 2-panel horizontal arrange (PWSH and CMD): a "quaked" one and a "standard" one. If your default profile is PWSH, then you'd write the next setting:

`"startupActions": "sp -V -p CMD -s .5 ; wt -w _quake \\; sp -V -p CMD -s .5"`

I hope someone finds it useful.

BTW, congratulations! WT is awesome and I really like where it is heading by becoming the Isildur's Bane of all terminals!  🥇

**Comment by dag03tsc** [comment]
@zadjii-msft, I list some additional properties from #10279 that could be useful inside each window entity, along with a short explanation.

    "tabs": {
      "name": string;
      "index": number;
      "panes": {
        "index": number;
        "profile": string;
        "size": number;
        "split": string;
        "splitMode": string;
      }[];
    }[];
    "defaultTab": "string";

The goal of `tabs` is to allow the auto-initialization of multiple tabs and panes and parameterize its behavior.

The property `name` should be any kind of unique identifier for the tab.
It'd mainly be used by the `defaultTab` property pairing process but it could also be used with a `switchToTab` action keybinding.

The property `index` must decide the order of creation, both for tabs and panes.

The goal of `defaultTab` is to allow the customization of the initially selected tab.
In other words, once all tabs have been initialized, automatically switch to the tab with `name` equal to `defaultTab`.
If `defaultTab` is undefined, fallback to the last or first one.

I think that `profile`, `size`, `split` and `splitMode` properties details are quite self-explanatory from the [Panes doc](https://docs.microsoft.com/en-us/windows/terminal/panes#customizing-panes-using-key-bindings) but I'll be glad to add its details too if needed.

**Comment by jhhcs** [comment]
Hey everyone, I found this thread while intending to open a feature request for the Quake mode window, which I will just leave here as a user story.

I am using the FAR manager inside Windows terminal and as a FAR user, I would like to be able to configure the Quake mode window so that it covers the entire screen. I currently have ConEmu setup in this exact way; it allows me to switch between using FAR and other windowed applications very quickly and efficiently.

**Comment by zadjii-msft** [comment]
@jhhcs FYI you don't need to use the `quakeMode` action to summon the window - you could always just configure the Terminal to launch in fullscreen mode, and rebind <kbd>win+\`</kbd> to a plain old [`globalSummon`](https://docs.microsoft.com/en-us/windows/terminal/customize-settings/actions#global-summon) action.

**Comment by jhhcs** [comment]
Oh thank you for that hint, yes that works quite nicely. I then discovered that toggling fullscreen with <kbd>Alt<kbd>+<kbd>Enter</kbd> can also be used to switch the Quake window to full screen as an alternative solution to my problem.

**Comment by rnett** [comment]
User story:
I'd like to be able to have one quake window for WSL ZSH, and another for Powershell.  This currently does not seem to be possible.

**Comment by zadjii-msft** [comment]
@rnett For now, you could always just give the two windows different names, and bind two separate `globalSummon` actions for those different names. You'd be able to get the dropdown on each now, but not necessarily the "snap to the top half" behavior.

**Comment by si-kotic** [comment]
With regards to tabs in Quake mode, that's what brought me here.  Unlike the OP of #11085 I would prefer them on the bottom of the window to the top.  Making this configurable would make sense to me, as would being able to toggle them on and off both as a default setting and also during a session.

**Comment by zadjii-msft** [comment]
@si-kotic you may also be interested in following https://github.com/microsoft/terminal/issues/835

**Comment by xorinzor** [comment]
Any updates on this? Having to toggle focus mode manually every time Quake mode starts is quite annoying.
Feels like this issue is trying to bundle too many separate issues into 1 fix.

Sure, profiles would be nice, but that's more of a "feature" then a bugfix.

**Comment by markusdd** [comment]
yes I am confused why I got redirected here.
This is a super simple request: launch a quake window with focus mode disabled. can't be done apparently, or does anyone have a solution?

I am not intesred in any of the other things being discussed, but the original ticket redirects here.

There is a -f switch to toggle focus mode on, why isn't there one to keep it off?

**Comment by zadjii-msft** [comment]
Alright, I'm calling it for this hackathon, but progress notes from the week:

Latest branch was: `dev/migrie/fhl-fall-2023/11162-quake-III-arena`

Sample JSON

```json
{
    "name": "_quake",
    "defaultProfile": "haunter.gif",
    "dockWindow": {
        "side": "bottom",
        "width": 0.3,
        "height": 0.75
    },
    "centerOnLaunch": true,
    "initialPosition": "",
}
```

* Moving the window settings into their own object is tedious, but easy. Almost everything becomes a per-window setting. 
* Layering window settings is _annoying_. They'll work like
  ```mermaid
  graph LR

  uq[user: _quake]
  uf[user: foo]
  ub[user: base]
  dq([defaults: _quake])
  db[defaults: base]

  uf  --> ub --> db
  uq --> dq --> ub
  ```
  - We make a "virtual" window settings blob if the user asks for a quake window.
  - If the user does specify settings for the `_quake` window, then it'll still be applied on top of the "builtin" quake settings
* If you don't use this at all, I don't think your settings will change at all[^1]
* [ ] I don't have hot-reloading working
* [ ] I don't have hot-renaming working
* [ ] I don't have startupActions working
* [ ] There are some other TODO!s I've left scattered.
* I've added a new settings blob to per-window settings: `dockWindow`. These are for #11162 and #11174
  - `side`: which side to "dock" to
  - `width` & `height`: How big to make the window. 
    - If >1, that value is treated as px
    - If <=1.0, that value is treated as a percent of the screen
  - It also combines with `centerOnLaunch`, to center the window on that side
  - [ ] It should warn if you set a value to <=0
  - [ ] It should maybe obey `initialPosition` if `centerOnLaunch` isn't set
  - [ ] It doesn't "dropdown" from the docked side 
  - [ ] It's a few pixels off from the right position. I think the `ncSize.height` is bigger than I'm expecting. 
* [ ] This altogether does nothing for #12594. It probably should!




[^1]: an empty `"windows":[]` may be added.

**Comment by drannex42** [comment]
Quick tip for everyone, you can set it to open in quake mode, and then set a keyboard shortcut to disable focus mode. I personally use win+ctrl+f for this. The tabs will remain in view for the entirety of the process lifetime.

**Comment by Squidyy** [comment]
> Quick tip for everyone, you can set it to open in quake mode, and then set a keyboard shortcut to disable focus mode. I personally use win+ctrl+f for this. The tabs will remain in view for the entirety of the process lifetime.

Thank you this is very helpful

**Comment by shookietea** [comment]
Leaving my question here because it's related- I'm wanting quake mode functionality with a maximized window.

I've got a `globalSummon` action setup for a separate named `_default` window, with focus and maximized arguments. This mostly works (without the quake animation). 

Is there a way to hide the taskbar icon when I've got this minimized, like the special `_quake` window does?

```json
{
  "command": {
    "action": "globalSummon",
    "desktop": "toCurrent",
    "MaximizeAfterToggle": true,
    "dropdownDuration": 0,
    "monitor": "toMouse",
    "name": "_default",
    "toggleVisibility": true
  },
  "keys": "win+`"
  }
```
`"startupActions": "wt --window _default -f -M \\; sp -V -s .666",`

**Comment by vladimmi** [comment]
How the hell does it take **THREE YEARS** for a basic feature and why does all "quake-style"-related issues are closed in favor of this? How hard is it to make already-existing mode actually usable and only after that slack however you want while people use it?

**Comment by zadjii-msft** [comment]
It's not really a "basic feature". It requires a pretty massive re-write of how our settings are stored. You can take a look at the all-up diff in https://github.com/microsoft/terminal/compare/main...dev/migrie/fhl-fall-2023/11162-quake-III-arena. There are piecemeal diffs listed in the OP as well. If you're passionate, you're more than free to take that over the finish line. 

You're also free to just use "globalSummon" instead of "quakeMode". IMO that's got quite a bit more flexibility. I honestly love focus mode, but don't care at all for the dropdown, so I just use `globalSummon`.

**Comment by shookietea** [comment]
> You're also free to just use "globalSummon" instead of "quakeMode". IMO that's got quite a bit more flexibility. I honestly love focus mode, but don't care at all for the dropdown, so I just use `globalSummon`. 

Is it currently possible to have "globalSummon" hide the taskbar icon when it is not visible?

**Comment by zadjii-msft** [comment]
With these two settings, I'm pretty sure it is:

* [`minimizeToNotificationArea`](https://learn.microsoft.com/en-us/windows/terminal/customize-settings/interaction#minimize-to-notification-area)
  * "when the terminal window is minimized, hide it (to the tray) instead"
* [`autoHideWindow`](https://learn.microsoft.com/en-us/windows/terminal/customize-settings/interaction#automatically-hide-on-focus-loss)
  * "hide the terminal window when it loses focus"

**Comment by shookietea** [comment]
Perfect! The animation is the lowest on my list, and wt is set up as I want now. Appreciate your support

**Comment by markusdd** [comment]
That really begs the question if quake mode should be dropped for the sake of that settings combination (meaning the setting is actually retained but that is what it does internally.)

So that I do not have to piuece this together myself, does anyone have a complete config snippet that emulates quake mode fully through global summon without stupid focus mode (that no one ever asked for, it's useless)

**Comment by shookietea** [comment]
Unfortunately, adding `"dropdownDuration": 200` doesn't work. When added to my command below, it causes the window size to be inconsistent after the first opening. It does add the animation, but breaks the behavior I want, which is full screen.

```json
    {
      "command": {
        "action": "globalSummon",
        "desktop": "toCurrent",
        "MaximizeAfterToggle": true,
        "monitor": "toMouse",
        "name": "_default",
        "toggleVisibility": true
      },
      "keys": "win+`"
    }
```

**Comment by Vampire** [comment]
@markusdd maybe you should learn to read ;-)
As I already said, just `"action": "globalSummon"` with all parameters default except for `"dropdownDuration": 200` is doing that, even the documentation of `quakeMode` tells you the `globalSummon` equivalent.

And regarding
> without stupid focus mode (that no one ever asked for, it's useless)

Just because **you** hate it and I don't like it, does not mean that no-one asked for it or that it is useless.
Even in this issue there is proof of love for the focus mode and without it, it is not anymore that quake-like that it is now.

**Comment by Fuzzyma** [comment]
I tried the quakemode equivalent of globalSummon and it doesnt have the quakebehavior if you dont give it the name _quake. But if you do, fockusmode is on again. Can we _please_ have this as a seperate ticket?

**Comment by amithegde** [comment]
any update on [#12150](https://github.com/microsoft/terminal/issues/12150)? It's been few years, it would be nice to have the ability to open in quake mode by default.

**Comment by isaacdontjelindell** [comment]
> I tried the quakemode equivalent of globalSummon and it doesnt have the quakebehavior if you dont give it the name _quake. But if you do, fockusmode is on again. Can we _please_ have this as a separate ticket?

I could not agree more here with @Fuzzyma. I'm seeing tickets requesting the ability to just turn off focus mode by default going back 3 years. Clearly the larger refactoring in this ticket isn't progressing, so what would it take to just add that single, special, setting?

**Comment by dag03tsc** [comment]
> any update on [#12150](https://github.com/microsoft/terminal/issues/12150)? It's been few years, it would be nice to have the ability to open in quake mode by default.

Please, check https://github.com/microsoft/terminal/issues/9992#issuecomment-851093521, because the `startupActions` property of each profile could be tuned to do exactly that with a value `wt -w _quake`. Do that with the default profile, and you're done, but I'm not sure whether there's a GUI for that or not, so bear in mind to apply that inside the Settings JSON with your favourite text editor.

**Comment by markusdd** [comment]
> > any update on [#12150](https://github.com/microsoft/terminal/issues/12150)? It's been few years, it would be nice to have the ability to open in quake mode by default.
> 
> [@amithegde](https://github.com/amithegde), please, check [#9992 (comment)](https://github.com/microsoft/terminal/issues/9992#issuecomment-851093521), because the `startupActions` setting could be tuned to do exactly that with a value `wt -w _quake`, but I'm not sure whether there's a GUI for that or not, so bear in mind to apply that inside the Settings JSON with your favourite text editor, and you're done.
> 
> [This is the current default Settings JSON](https://github.com/microsoft/terminal/blob/main/src/cascadia/TerminalSettingsModel/defaults.json#L32), where you may find at which level the property should lie, in case you find it missing inside your local version of the JSON.

But does this approach also allow it to launch in quake mode with the damn focus mode turned off by default? Like if anyone wants it fine, but to 90%+ it's just a nuisance to not see the tabs one has open...and its an extra action each time to turn it off...

**Comment by dag03tsc** [comment]
> > > any update on [#12150](https://github.com/microsoft/terminal/issues/12150)? It's been few years, it would be nice to have the ability to open in quake mode by default.
> > 
> > 
> > [@amithegde](https://github.com/amithegde), please, check [#9992 (comment)](https://github.com/microsoft/terminal/issues/9992#issuecomment-851093521), because the `startupActions` setting could be tuned to do exactly that with a value `wt -w _quake`, but I'm not sure whether there's a GUI for that or not, so bear in mind to apply that inside the Settings JSON with your favourite text editor, and you're done.
> > [This is the current default Settings JSON](https://github.com/microsoft/terminal/blob/main/src/cascadia/TerminalSettingsModel/defaults.json#L32), where you may find at which level the property should lie, in case you find it missing inside your local version of the JSON.
> 
> But does this approach also allow it to launch in quake mode with the damn focus mode turned off by default? Like if anyone wants it fine, but to 90%+ it's just a nuisance to not see the tabs one has open...and its an extra action each time to turn it off...

It does not, but that is actually expected by definition; quoting [Mike's previous comment](https://github.com/microsoft/terminal/issues/11085#issuecomment-970662470), the `_quake` window is a kind of special window with properties `launchMode:focus`, `snapTo:top`, and `minimizeToTray:true` set.

Maybe you do not want the `_quake` window, but a default window with properties `launchMode:maximized`, `snapTo:top`, and `minimizeToTray:true` set, so you may add those manually inside the `settings.json` file to build your own flavoured default window design. If you want to apply that settings with different values for each window rather than as a default, well, that's the purpose of this thread.😉

Regarding the opened tabs list, it shows with the `Ctrl+Tab` shortcut, and the focus mode keeps tabs navigation active, which is handy, though I prefer panes over tabs.

I know it's a matter of taste for sure, but I find the extra space of both the focus and the fullscreen modes really inmmersive and useful indeed when combined with some additional properties and shortcuts. It's the closest experience to classical getty virtual consoles, namely `Ctrl+Alt+[F1-F6]`, and even improved, thanks to dynamic panes, summoning shortcuts, flexiblity of consoles support, etc.

**Comment by Fuzzyma** [comment]
@dag03tsc yeah no, I tried all that and it does not behave like the quake window. This discussion is ridiculous. Adding a flag for focusMode in quake would probably take 10min for someone knowing the code. But instead every now and then someone new comes along and repeats half baked solutions. That doesn't help anyone and only adds to the frustration.
